### PR TITLE
130 generate batch data from transform

### LINKED ISSUE #130: generate batch data from 1 big image
**Is your feature request related to a problem? Please describe.**
As many medical images are big and hard to label(especially segmentation), most of the medical training datasets contain only 10 - 100 big images. We need to crop 1 big image to generate a batch of smaller images for training.

**Describe the solution you'd like**
There are many challenges in this task:
1. After cropping, the output data shape will be [N, C, H, W, D] instead of original [C, H, W, D].
    And still need to apply other transforms to every output data.
2. In the DataLoader, got a batch of data in [N, C, H, W, D], need to reshape to [batch x N, C, H, W, D].
3. Need to support many different methods to crop expected small batch:
    (1) Use the sliding window to crop image & label together.
    (2) Randomly crop image & label based on positive/negative ratio(use label field).
    ... ...

**Describe alternatives you've considered**
In TorchVsion's vanilla transforms, there is a very basic reference method:
```py
transform = Compose([
    TenCrop(size), # this is a list of PIL Images
    # returns a 4D tensor
    Lambda(lambda crops: torch.stack([ToTensor()(crop) for crop in crops]))
])
```

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #130 .

### Description
This PR mainly contains 4 features:
1. **Generate 1 batch** data from crop transform, for example:
    input 3 images, crop out 4 smaller images on every original images, then generate batch=12.
    All the following transforms can still apply to every item in the batch list(_Compose function_).
2. **General crop** transform at the vanilla level.
    It's a general crop transform, can support "center + size" or "start point + end point" modes.
3. Randomly crop out batch data based on a **positive/negative label ratio**.
    According to the label values, randomly select "crop centers" at positive or negative label positions.
4. Enhancement to default **collate_fn** for **list** data type.

Usage example:
1. Define transforms for dict data:
```py
transforms = [
    LoadNiftid(keys=['image', 'extral', 'label']),
    NormalizeIntensityd(keys=['image'], divdior=1.0),
    RandCropByPosNegLabeld(
        keys=['image', 'extral'],
        label_key='label',
        size=[128, 128, 128],
        pos=1, neg=1, num_samples=2
    ),
    RandRotate90d(keys=['image', 'label'])
    ... ...
]
```
2. Define Dataset, DataLoader and use "list_data_collate":
```py
dataset = NiftiDataset(transforms=transforms)
dataloader = DataLoader(dataset=dataset, batch_size=4, collate_fn=list_data_collate)
```

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @ericspod , @atbenmurray , @wyli  and @yanchengnv ,

Could you please help review this PR?
As we still don't have an example of dict level transforms, I put the usage in the PR description directly.
I will develop a training example when all the dict level transforms & training example are ready.
Thanks.

File: monai/transforms/composables.py [hunk 1]
```diff
@@ -120,6 +121,61 @@ def __call__(self, data):
...
+        assert batch_size >= 0, 'batch_size must be greater than or equal to 0.'
+        self.label_key = label_key
+        self.size = size
+        self.pos_ratio = float(pos) / (float(pos) + float(neg))
+        self.batch_size = batch_size
+
+    def __call__(self, data):
+        d = dict(data)
+        label = d[self.label_key]
+        centers = generate_pos_neg_label_crop_centers(label, self.size, self.batch_size, self.pos_ratio, self.R)
```
Comment (wyli) [inline monai/transforms/composables.py]: would be great to keep a record of the `centers` after generating the patches, i think we could make `centers` an instance variable, or make new keys in the output data

Comment (Nic-Ma) [reply monai/transforms/composables.py]: Hi @wyli ,

Thanks for your comments, I already updated code to save centers and use randomize API.

**Comment by Nic-Ma** [comment]
Thanks @wyli for the review and comments.
I also added another test case to testCompose and added detailed docstring to Compose.
Do you guys have any more comments?
Thanks.

File: monai/transforms/compose.py [hunk 2]
```diff
@@ -86,7 +99,13 @@ def __init__(self, transforms=None):
 
     def __call__(self, input_):
         for transform in self.transforms:
-            input_ = transform(input_)
+            # if some transform generated batch list of data in the transform chain,
+            # all the following transforms should apply to every item of the list.
+            if isinstance(input_, list):
+                for i, item in enumerate(input_):
+                    input_[i] = transform(item)
```
Comment (wyli) [inline monai/transforms/compose.py]: potentially losing the capability of processing `enumerate(input_)` with multiple workers... personally I don't quite like this idea of treating the image-patch sampling as a composable transform. But I guess I can approve this PR because it doesn't break any existing features

File: monai/transforms/transforms.py [hunk 3]
```diff
@@ -206,6 +206,58 @@ def __call__(self, img):
         return rotator(img)
 
 
+@export
+class GeneralCrop:
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/transforms.py]: Fixed in latest code.

File: monai/transforms/transforms.py [hunk 4]
```diff
@@ -206,6 +206,58 @@ def __call__(self, img):
...
+            self.roi_end = roi_end
+
+    def __call__(self, img):
+        max_end = img.shape[1:]
+        assert (np.subtract(max_end, self.roi_start) >= 0).all(), 'roi start out of image space.'
+        assert (np.subtract(max_end, self.roi_end) >= 0).all(), 'roi end out of image space.'
+        assert (np.subtract(self.roi_end, self.roi_start) >= 0).all(), 'invalid roi range.'
+        if len(self.roi_start) == 1:
+            return img[:, self.roi_start[0]:self.roi_end[0]].copy()
+        elif len(self.roi_start) == 2:
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/transforms.py]: Fixed in latest code.

File: monai/transforms/composables.py [hunk 5]
```diff
@@ -120,6 +121,65 @@ def __call__(self, data):
...
+        label_key (str): name of key for label image, this will be used for finding foreground/background.
+        size (list, tuple): the size of the crop region e.g. [224,224,128]
+        pos (int, float): used to calculate the ratio ``pos / (pos + neg)`` for the probability to pick a
+          foreground voxel as a center rather than a background voxel.
+        neg (int, float): used to calculate the ratio ``pos / (pos + neg)`` for the probability to pick a
+          foreground voxel as a center rather than a background voxel.
+        batch_size (int): number of samples (crop regions) to take in each batch.
+    """
+
+    def __init__(self, keys, label_key, size, pos=1, neg=1, batch_size=1):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/composables.py]: Fixed in latest code.

File: monai/transforms/utils.py [hunk 6]
```diff
@@ -157,3 +157,54 @@ def one_hot(labels, num_classes):
...
+    max_size = label.shape[1:]
+    assert len(max_size) == len(size), 'expected size does not match label dim.'
+    assert (np.subtract(max_size, size) >= 0).all(), 'proposed roi is larger than image itself.'
+
+    # Select subregion to assure valid roi
+    valid_start = np.floor_divide(size, 2)
+    valid_end = np.subtract(max_size + np.array(1), size / np.array(2)).astype(np.uint16)  # add 1 for random
+    # int generation to have full range on upper side, but subtract unfloored size/2 to prevent rounded range
+    # from being too high
+    for i in range(2):  # need this because np.random.randint does not work with same start and end
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/utils.py]: Fixed in latest code.

File: monai/transforms/utils.py [hunk 7]
```diff
@@ -157,3 +157,54 @@ def one_hot(labels, num_classes):
     onehot = y[labels.flatten()]
 
     return onehot.reshape(tuple(labels.shape) + (num_classes,)).astype(labels.dtype)
+
+
+def generate_pos_neg_label_crop_centers(label, size, batch_size, pos_ratio, rand_state):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/utils.py]: Fixed in latest code.

[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/utils.py]: Fixed in latest code.

File: monai/transforms/compose.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/transforms/compose.py]: Thanks for your review, I already updated code for all comments.

File: monai/transforms/composables.py [hunk 8]
```diff
@@ -120,6 +121,66 @@ def __call__(self, data):
...
+
+    def __call__(self, data):
+        d = dict(data)
+        label = d[self.label_key]
+        self.randomise(label)
+        results = [dict()] * self.num_samples
+        for key in data.keys():
+            if key in self.keys:
+                img = d[key]
+                image_samples = list()
```
Comment (wyli) [inline monai/transforms/composables.py]: this is unused...

File: monai/transforms/composables.py [hunk 9]
```diff
@@ -120,6 +121,66 @@ def __call__(self, data):
...
+                    results[i][key] = cropper(img)
+            else:
+                for i in range(self.num_samples):
+                    results[i][key] = data[key]
+
+        # also need to crop label field
+        for i, center in enumerate(self.centers):
+            cropper = SpatialCrop(roi_center=tuple(center), roi_size=self.size)
+            results[i][self.label_key] = cropper(label)
+        return results
```
Comment (wyli) [inline monai/transforms/composables.py]: in this case the user couldn't choose to not cropping data in `self.label_key`?  I suppose `self.keys` already have everything we want to crop

File: monai/transforms/composables.py (hunk 8 above)
Comment (Nic-Ma) [reply monai/transforms/composables.py]: Removed in the latest patch.
Thanks.

File: monai/transforms/composables.py (hunk 9 above)
Comment (Nic-Ma) [reply monai/transforms/composables.py]: Adjusted the logic and unit test in the latest patch.
Thanks.
